circuit-codegen-annotations/src/main/kotlin/com/slack/circuit/codegen/annotations/CircuitInject.kt

```diff
@@ -0,0 +1,139 @@
+ *
+ * // Generates
+ * @ContributesMultibinding(AppScope::class)
+ * class HomePresenterFactory @Inject constructor() : Presenter.Factory { ... }
+ * ```
+ *
+ * **UI**
+ * ```kotlin
+ * @CircuitInject(HomeScreen::class, AppScope::class)
+ * class HomeUi @Inject constructor(...) : Ui<HomeState> { ... }
+ *
+ * // Generates
+ * @ContributesMultibinding(AppScope::class)
+ * class HomeUiFactory @Inject constructor() : Ui.Factory { ... }
+ * ```
+ *
+ * ## Functions
+ *
+ * Simple functions can be annotated and have a corresponding [Presenter.Factory] generated. This is
+ * primarily useful for simple cases where a class is just technical tedium.
+ *
+ * **Requirements**
+ * - Presenter function names _must_ end in `Presenter`, otherwise they will be treated as UI
+ * functions.
+ * - Presenter functions _must_ return a [CircuitUiState] type.
+ * - UI functions can optionally accept a [CircuitUiState] type as a parameter, but it is not
+ * required.
+ * - UI functions _must_ return [Unit].
+ * - Both presenter and UI functions _must_ be [Composable].
+ *
+ * **Presenter**
+ * ```kotlin
+ * @CircuitInject(HomeScreen::class, AppScope::class)
+ * @Composable
+ * fun HomePresenter(): HomeState { ... }
+ *
+ * // Generates
+ * @ContributesMultibinding(AppScope::class)
+ * class HomePresenterFactory @Inject constructor() : Presenter.Factory { ... }
+ * ```
+ *
+ * **UI**
+ * ```kotlin
+ * @CircuitInject(HomeScreen::class, AppScope::class)
+ * @Composable
+ * fun Home(state: HomeState) { ... }
+ *
+ * // Generates
+ * @ContributesMultibinding(AppScope::class)
+ * class HomeUiFactory @Inject constructor() : Ui.Factory { ... }
+ * ```
+ *
+ * ## Assisted injection
+ *
+ * Any type that is offered in [Presenter.Factory] and [Ui.Factory] can be offered as an assisted
+ * injection to types using Dagger [AssistedInject]. For these cases, the [AssistedFactory]
+ * -annotated interface should be annotated with [CircuitInject] instead of the enclosing class.
+ *
+ * Types available for assisted injection are:
+ * - [Screen] – the screen key used to create the [Presenter] or [Ui].
+ * - [Navigator] – (presenters only)
+ * - [CircuitConfig]
+ *
+ * Each should only be defined at-most once.
+ *
+ * **Examples**
+ * ```kotlin
+ * // Function example
+ * @CircuitInject(HomeScreen::class, AppScope::class)
+ * @Composable
+ * fun HomePresenter(screen: Screen, navigator: Navigator): HomeState { ... }
+ *
+ * // Class example
+ * class HomePresenter @AssistedInject constructor(
+ *   @Assisted screen: Screen,
+ *   @Assisted navigator: Navigator,
+ *   ...
+ * ) : Presenter<HomeState> {
+ *   // ...
+ *
+ *   @CircuitInject(HomeScreen::class, AppScope::class)
+ *   @AssistedFactory
+ *   fun interface Factory {
+ *     fun create(screen: Screen, navigator: Navigator): HomePresenter
+ *   }
+ * }
+ * ```
+ */
+@Target(AnnotationTarget.CLASS, AnnotationTarget.FUNCTION)
+public annotation class CircuitInject(
+  val screen: KClass<out Screen>,
+  val scope: KClass<*>,
```

Comment:
why can't the scope just be of type  `KClass.` What is the purpose of the generic here?

Reply:
You can't write raw types in kotlin like you can in Java. For example, in Java it's legal to write `List foo = ...`, but in Kotlin you _must_ write the generic, even if it's just a star. `val foo: List = ...` isn't legal syntax